github-extension/src/browser/github-quick-open-service.ts

```diff
@@ -0,0 +1,300 @@
+import { StatusBar, StatusBarAlignment } from '@theia/core/lib/browser/status-bar/status-bar';
+
+@injectable()
+export class GitHubQuickOpenService {
+
+    private rootUri: URI | undefined;
+    private gitUrlRegex = /(?:git|ssh|https?|git@[-\w.]+):(\/\/)?(.*?)(\.git)(\/?|#[-\d\w._]+?)$/;
+
+    constructor(
+        @inject(MessageService) protected readonly messageService: MessageService,
+        @inject(QuickOpenService) protected readonly quickOpenService: QuickOpenService,
+        @inject(OpenerService) protected readonly openerService: OpenerService,
+        @inject(GitHubProperties) protected readonly gitHubProperties: GitHubProperties,
+        @inject(GithubService) protected readonly gitHubService: GithubService,
+        @inject(Git) protected readonly git: Git,
+        @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService,
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(SelectionService) protected readonly selectionService: SelectionService,
+        @inject(StatusBar) protected readonly statusBar: StatusBar
+    ) {
+        workspaceService.root.then(root => {
+            if (root) {
+                this.rootUri = new URI(root.uri);
+            }
+        });
+    }
+
+    openConfigurationFile() {
+        open(this.openerService, propertiesUri);
+    }
+
+    async cloneRepository(): Promise<void> {
+        const __this = this;
+
+        let pageNumber = 1;
+        let pageSize = 100;
+
+        let user = await this.gitHubService.getCurrentUser(this.gitHubProperties.getProperties());
+        let organizations = await this.gitHubService.getOrganizations(pageNumber, pageSize, this.gitHubProperties.getProperties());
+
+        enum CLONE_PROGRESS {
+            START,
+            STOP
+        }
+
+        const onCloningProgress = (progress: CLONE_PROGRESS, projectName?: string) => {
+            if (progress == CLONE_PROGRESS.START) {
+                this.statusBar.setElement("github-cloning", {text: "$(download) Cloning: " + projectName, alignment: StatusBarAlignment.LEFT});
+            } else if (progress == CLONE_PROGRESS.STOP) {
+                this.statusBar.removeElement("github-cloning");
+            }
+        };
+
+        const cloneRepository = (item: GitHubQuickOpenItem<Repository>) => {
+            if (!__this.rootUri) {
+                return;
+            }
+
+            onCloningProgress(CLONE_PROGRESS.START, item.ref.name);
+
+            let localUri = __this.rootUri;
+
+            __this.getDirectory(localUri).then(() => {
+                __this.fileSystem.createFolder(localUri.resolve(item.ref.name).toString()).then(dumbFolder => {
+                    __this.git.clone(item.ref.html_url, {localUri: dumbFolder.uri}).then(() => {
+                        onCloningProgress(CLONE_PROGRESS.STOP);
+                        __this.messageService.info(`Repository ${item.ref.name} successfully cloned.`);
```

Comment:
would it be possible to rewrite such nested blocks with the `await` operator?

Reply:
Hmm, need to try this

Reply:
Maybe it'll allow getting rid of introducing the `__this` variable.